_posts/2020-12-01-the-road-to-the-jit.md

```diff
@@ -0,0 +1,382 @@
+code. In Lukas's prototype, the native code template for each
+instruction was contained in text files similar to the other files
+used by the loader.  That was inflexible, so it was decided to use
+some library that could generate native code.  While some pure C
+libraries could have been used, the C++ library AsmJIT was more
+convenient in practical use than any of the C libraries.  Therefore
+the part of the loader that translates BEAM instructions to native
+code needed to be written in C++, but the rest of the runtime system
+is still pure C code and will remain so.
+
+John joined the practical work on the rejigged JIT project at the end
+of March.
+
+On April 7, 2020, John reached the "prompt beer" milestone.
+
+### Prompt Beer
+
+When the Erlang system is started, a surprisingly large amount of code is
+executed before the prompt appears. On the one hand that means that
+the translation of many instructions needs to be implemented
+before it would be possible to even start the Erlang system, let alone run
+any test suites or benchmarks.
+
+On the other hand, when the prompt finally appears, it is a major
+milestone worth celebrating with some prompt beer or other appropriate
+beverage or by taking the rest of the evening off.
+
+### Maturing the new JIT
+
+On April 14 John got Dialyzer running with the JIT, and on April 17
+after some improvements to the code generation Dialyzer was only about
+10 percent slower with the JIT than with HiPE. (At the time of
+writing, Dialyzer runs roughly as fast with the JIT as it did with
+HiPE, although it has been become increasingly difficult to do a fair
+comparison since HiPE doesn't work beyond OTP 23.)
+
+The next major milestone was reached May 6 when line numbers in
+stack traces were implemented. That meant that many more test cases now
+succeeded.
+
+Soon after that, all test suites could be run successfully.  It was decided
+to develop the new JIT into a product that could be included in OTP 24.  During
+the summer and early fall, the following was done:
+
+* A major refactoring of the BEAM loader so that as much code as possible
+  could be shared between the JIT and the BEAM interpreter. (The BEAM interpreter
+  is only used on platforms that don't support the JIT.)
+
+* Implementing and polishing of important but less used features such
+  as tracing, save calls (see the flag `save_calls` for
+  [process_flag/2][process_flag]), and [perf][perf] support.
+
+* Shrinking of the code size of the generated native code.
+
+* Porting the JIT to Windows (about three weeks of work).
+
+* Making it possible to use the native stack pointer register and
+  stack manipulation instructions. That improved perf support and
+  slightly reduced the size of the native code.
+
+The work culminated in a public [pull request][beamasm] that Lukas
+created during his [presentation][jit_presentation] of the new JIT
+during his presentation in the [virtual Code BEAM][CodeBEAM]
+conference on September 11.
+
```

Comment:
"The work culminated in a public [pull request][beamasm] that Lukas
created during his [presentation][jit_presentation] of the new JIT
during his presentation in the [virtual Code BEAM][CodeBEAM]
conference on September 11." =>

"The work culminated in a public [pull request][beamasm] that Lukas
created during his [virtual Code BEAM][CodeBEAM] [presentation][jit_presentation] of the new JIT on September 11."